File: go/arrow/array/data_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package array

import (
	"testing"

	"github.com/apache/arrow/go/v16/arrow"
	"github.com/apache/arrow/go/v16/arrow/memory"
	"github.com/stretchr/testify/assert"
)

func TestDataReset(t *testing.T) {
	var (
		buffers1 = make([]*memory.Buffer, 0, 3)
		buffers2 = make([]*memory.Buffer, 0, 3)
	)
	for i := 0; i < cap(buffers1); i++ {
		buffers1 = append(buffers1, memory.NewBufferBytes([]byte("some-bytes1")))
		buffers2 = append(buffers2, memory.NewBufferBytes([]byte("some-bytes2")))
	}

	data := NewData(&arrow.StringType{}, 10, buffers1, nil, 0, 0)
	data.Reset(&arrow.Int64Type{}, 5, buffers2, nil, 1, 2)

	for i := 0; i < 2; i++ {
		assert.Equal(t, buffers2, data.Buffers())
		assert.Equal(t, &arrow.Int64Type{}, data.DataType())
		assert.Equal(t, 1, data.NullN())
		assert.Equal(t, 2, data.Offset())
		assert.Equal(t, 5, data.Len())

		// Make sure it works when resetting the data with its own buffers (new buffers are retained
		// before old ones are released.)
		data.Reset(&arrow.Int64Type{}, 5, data.Buffers(), nil, 1, 2)
	}
}

func TestSizeInBytes(t *testing.T) {
	var buffers1 = make([]*memory.Buffer, 0, 3)

	for i := 0; i < cap(buffers1); i++ {
		buffers1 = append(buffers1, memory.NewBufferBytes([]byte("15-bytes-buffer")))
	}
	data := NewData(&arrow.StringType{}, 10, buffers1, nil, 0, 0)
	var arrayData arrow.ArrayData = data
	dataWithChild := NewData(&arrow.StringType{}, 10, buffers1, []arrow.ArrayData{arrayData}, 0, 0)

	t.Run("buffers only", func(t *testing.T) {
		expectedSize := uint64(45)
		if actualSize := data.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})

	t.Run("buffers and child data", func(t *testing.T) {
		// 45 bytes in buffers, 45 bytes in child data
		expectedSize := uint64(90)
		if actualSize := dataWithChild.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})

	t.Run("buffers and nested child data", func(t *testing.T) {
		var dataWithChildArrayData arrow.ArrayData = dataWithChild
		var dataWithNestedChild arrow.ArrayData = NewData(&arrow.StringType{}, 10, buffers1, []arrow.ArrayData{dataWithChildArrayData}, 0, 0)
		// 45 bytes in buffers, 90 bytes in nested child data
		expectedSize := uint64(135)
		if actualSize := dataWithNestedChild.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})

	t.Run("buffers and dictionary", func(t *testing.T) {
		dictData := data
		dataWithDict := NewDataWithDictionary(&arrow.StringType{}, 10, buffers1, 0, 0, dictData)
		// 45 bytes in buffers, 45 bytes in dictionary
		expectedSize := uint64(90)
		if actualSize := dataWithDict.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})

	t.Run("sliced data", func(t *testing.T) {
		sliceData := NewSliceData(arrayData, 3, 5)
		// offset is not taken into account in SizeInBytes()
		expectedSize := uint64(45)
		if actualSize := sliceData.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})

	t.Run("sliced data with children", func(t *testing.T) {
		var dataWithChildArrayData arrow.ArrayData = dataWithChild
		sliceData := NewSliceData(dataWithChildArrayData, 3, 5)
		// offset is not taken into account in SizeInBytes()
		expectedSize := uint64(90)
		if actualSize := sliceData.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})

	t.Run("buffers with children which are sliced data", func(t *testing.T) {
		sliceData := NewSliceData(arrayData, 3, 5)
		dataWithSlicedChildren := NewData(&arrow.StringType{}, 10, buffers1, []arrow.ArrayData{sliceData}, 0, 0)
		// offset is not taken into account in SizeInBytes()
		expectedSize := uint64(90)
		if actualSize := dataWithSlicedChildren.SizeInBytes(); actualSize != expectedSize {
			t.Errorf("expected size %d, got %d", expectedSize, actualSize)
		}
	})
}

File: cpp/src/arrow/array/data_test.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <type_traits>

#include <gtest/gtest.h>

#include "arrow/array/array_base.h"
#include "arrow/array/data.h"
#include "arrow/testing/gtest_util.h"

namespace arrow {

TEST(ArraySpan, SetSlice) {
  auto arr = ArrayFromJSON(int32(), "[0, 1, 2, 3, 4, 5, 6, null, 7, 8, 9]");
  ArraySpan span(*arr->data());
  ASSERT_EQ(span.length, arr->length());
  ASSERT_EQ(span.null_count, 1);
  ASSERT_EQ(span.offset, 0);

  span.SetSlice(0, 7);
  ASSERT_EQ(span.length, 7);
  ASSERT_EQ(span.null_count, kUnknownNullCount);
  ASSERT_EQ(span.offset, 0);
  ASSERT_EQ(span.GetNullCount(), 0);

  span.SetSlice(7, 4);
  ASSERT_EQ(span.length, 4);
  ASSERT_EQ(span.null_count, kUnknownNullCount);
  ASSERT_EQ(span.offset, 7);
  ASSERT_EQ(span.GetNullCount(), 1);
}

TEST(ArrayData, GetSpanRespectsOffset) {
  std::vector<uint16_t> values = {1, 2, 3, 4, 5};

  auto buffer = Buffer::FromVector(std::move(values));
  auto data = ArrayData::Make(uint16(), /*length=*/3, {nullptr, buffer}, /*null_count=*/0,
                              /*offset=*/1);
  auto span = data->GetSpan<uint16_t>(1, 3);

  const bool is_const_pointer = std::is_same_v<decltype(span)::pointer, const uint16_t*>;
  ASSERT_TRUE(is_const_pointer);

  EXPECT_EQ(span.size(), 3);
  EXPECT_EQ(span[0], 2);
  EXPECT_EQ(span[1], 3);
  EXPECT_EQ(span[2], 4);
}

TEST(ArrayData, GetMutableSpanRespectsOffset) {
  std::vector<uint16_t> values = {10, 20, 30, 40, 50};

  auto buffer = std::make_shared<MutableBuffer>(reinterpret_cast<uint8_t*>(values.data()),
                                                values.size() * sizeof(uint16_t));
  std::vector<std::shared_ptr<Buffer>> buffers = {nullptr, buffer};

  auto data =
      ArrayData::Make(uint16(), /*length=*/3, buffers, /*null_count=*/0, /*offset=*/1);
  auto span = data->GetMutableSpan<uint16_t>(1, 3);

  const bool is_mut_pointer = std::is_same_v<decltype(span)::pointer, uint16_t*>;
  ASSERT_TRUE(is_mut_pointer);

  EXPECT_EQ(span.size(), 3);
  EXPECT_EQ(span[0], 20);
  EXPECT_EQ(span[1], 30);
  EXPECT_EQ(span[2], 40);

  span[0] = 200;
  span[1] = 300;
  span[2] = 400;

  auto raw = reinterpret_cast<uint16_t*>(buffer->mutable_data());

  EXPECT_EQ(raw[1], 200);
  EXPECT_EQ(raw[2], 300);
  EXPECT_EQ(raw[3], 400);
}

TEST(ArrayData, GetSpanNullBuffer) {
  auto data = ArrayData::Make(uint16(), /*length=*/0, /*buffers=*/{nullptr, nullptr},
                              /*null_count=*/0, /*offset=*/0);
  auto span = data->GetSpan<uint16_t>(1, 0);
  EXPECT_EQ(span.size(), 0);
}

TEST(ArrayData, GetMutableSpanNullBuffer) {
  auto data = ArrayData::Make(uint16(), /*length=*/0, /*buffers=*/{nullptr, nullptr},
                              /*null_count=*/0, /*offset=*/0);
  auto span = data->GetMutableSpan<uint16_t>(1, 0);
  EXPECT_EQ(span.size(), 0);
}

}  // namespace arrow
